src/Infrastructure.VS/ActiveDocumentLocator.cs

```diff
@@ -40,22 +42,24 @@ public interface IActiveDocumentLocator
     internal class ActiveDocumentLocator : IActiveDocumentLocator
     {
         private readonly ITextDocumentProvider textDocumentProvider;
+        private readonly IThreadHandling threadHandling;
         private IVsMonitorSelection monitorSelection;
 
         [ImportingConstructor]
-        public ActiveDocumentLocator([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, ITextDocumentProvider textDocumentProvider)
+        public ActiveDocumentLocator([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, ITextDocumentProvider textDocumentProvider, IThreadHandling threadHandling)
         {
             this.textDocumentProvider = textDocumentProvider;
-            
-            RunOnUIThread.Run(() =>
+            this.threadHandling = threadHandling;
+
+            threadHandling.RunOnUIThread(() =>
```

Comment:
Looking again at this code, I think running it asynchronically is not the right way to go. I think it's better to do "throwIfNotOnUiThread" and then simply call it, because we expect all MEF ctors to be called on the UI thread. WDYT? cc @duncanp-sonar

Reply:
@rita-sonarsource I _thought_ that was the case, but I'm no longer sure. I can't remember if it was something I read or something noticed when debugging. I'll have a look.

Reply:
From the [VS Threading Cookbook](https://github.com/microsoft/vs-threading/blob/main/doc/cookbook_vs.md#how-do-i-effectively-verify-that-my-code-is-fully-free-threaded):

> Code is free-threaded if it... can complete on the caller's thread, no matter which thread that is.
Important places to be free-threaded include:
> 1. Code that runs in the MEF activation code paths (e.g. importing constructors, OnImportsSatisfied callbacks, and any code called therefrom).


So MEF constructors should not assume that they are called on the main thread.